ironfish-rust/src/errors.rs

```diff
@@ -2,19 +2,26 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
 
+use std::backtrace::Backtrace;
 use std::error::Error;
 use std::fmt;
 use std::io;
 use std::num;
 use std::string;
 
+#[derive(Debug)]
+pub struct IronfishError {
+    pub kind: IronfishErrorKind,
+    pub backtrace: Backtrace,
+}
+
 /// Error type to handle all errors within the code and dependency-raised
 /// errors. This serves 2 purposes. The first is to keep a consistent error type
 /// in the code to reduce the cognitive load needed for using Result and Error
 /// types. The second is to give a singular type to convert into NAPI errors to
 /// be raised on the Javascript side.
 #[derive(Debug)]
-pub enum IronfishError {
+pub enum IronfishErrorKind {
     BellpersonSynthesis(bellperson::SynthesisError),
```

Comment:
Usually the inner error is not put into the "ErrorKind" enum, but inside the "Error" struct instead. This is for example what happens with `io::Error`, where its `io::ErrorKind` is a simple enumeration that does not hold any data.

Now I know that mimicking `io::Error` would make this change much more large, but I wanted to point it out in case you want to consider it.

Reply:
Thanks for the heads up, I will look into this.

Reply:
I looked into this and I am not sure how this is different than the `io::Error` type, the definition of the enum is not in the struct. I also see we construct an error the same way already in our codebase:
https://github.com/iron-fish/ironfish/blob/cbd402cb6cf685a090082bb6babfb8e4d513e501/ironfish-phase2/src/lib.rs#L1133-L1135

Reply:
Compare:

```rust
Error::new(ErrorKind::Other, inner_error)
IronFishError::new(IronFishErrorKind::BellpersonSytnesis(inner_error))
```

`io::Error` is equivalent to a 2-tuple `(io::ErrorKind, Box<dyn Error>)`. In particular, `io::ErrorKind` does not hold any information about the inner error, it's just a simple enumeration that does not contain any data.

With the current code in this PR we have the other way around: `IronFishError` is equivalent to a 1-tuple `(IronFishErrorKind)` (+ the backtrace, but that's not relevant for this discussion), and `IronFishError` does not directly contain information about the inner error. `IronFishErrorKind` instead may hold the inner error, and it's not a simple enumeration, but a more complex one.

Reply:
Ahhh I see! I was just consider the plain enumerable errors. I will look into changing to be more idiomatic

Reply:
This might require a lot of changes, so do it only if you really want to do it :)

Reply:
It actually wasn't that bad, take a look at the newest version of the PR and let me know what you think. Feel free to review there or here, github unfortunately archived this comment.